onnxruntime/test/onnx/callables.h

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include <type_traits>
+#include <tuple>
+
+namespace onnxruntime {
+namespace test {
+
+/// <summary>
+/// This class serves as a callable entity that can be used
+/// either in a threadpool or as a callback fro async operations. It supports both
+/// plain functions as well as member functions and provides type
+/// erasure functionality. Unlike std::function, it
+/// does do any memory allocations and it is cheap to copy.
+/// </summary>
+/// 
+/// <example>
+/// Plain function:
+///   int g(void* context, float);
+///   Callable<int, float> cb(nullptr, g);
+///   int result = cb.Invoke(1.2f);
+///
+/// Class method:
+/// class A {
+///    int f(float);
+/// };
+/// 
+/// A a;
+/// CallableFactory<A, int, float> f(&a);
+/// - We get back Callable<int, float> as above but call instance method
+/// auto cb = f.GetCallable<&A::f>();
+/// int result = cb.Invoke(1.2f);
+/// </example>
+/// 
+/// <typeparam name="Result">Call back result type</typeparam>
+/// <typeparam name="...Args">Types of args for the callable</typeparam>
+template <typename    Result,
+          typename... Args>
+class Callable {
+public:
+
+  using
+  ContextType = void;
+
+  using
+  ResultType = Result;
+
+  using
+  FunctionType = ResultType(*)(ContextType*, Args...);
+
+  // You can use this directly if you have a simple
+  // function like C-function or a stateless lambda
+  // For methods we want to make use of AsyncMethodCallback
+  // And covert it to a Callable
+  Callable(ContextType*  context,
+    FunctionType func) :
+    context_(context),
+    func_(func) {
+  }
+
+  Callable() : Callable(nullptr, nullptr) {}
+
+  Callable(const Callable&) = default;
+  Callable& operator=(const Callable&) = default;
+
+  Callable(Callable&&) = default;
+  Callable& operator=(Callable&&) = default;
+
+  void Clear() {
+    context_ = nullptr;
+    func_ = nullptr;
+  }
+
+  operator bool() const { return func_ != nullptr; }
+
+  bool Valid() const { return func_ != nullptr; }
+
+  ContextType* GetContext() const { return context_; }
+
+  FunctionType GetFunctionPtr() const { return func_; }
+
+  // Differentiate instantiation on the presence
+  // of return value. The below two overloads
+  // must be mutually exclusive
+  template<typename T = ResultType>
+  typename std::enable_if<!std::is_same<void, T>::value, T>::type
+  Invoke(Args... args) const {
+    return func_(context_, std::forward<Args>(args)...);
```

Comment:
>func_ [](start = 11, length = 5)

If by accident func_ is nullptr and user forget to check with Valid() this will result in segfault?

Reply:
In async contexts we are trying to minimize the number of exceptions thrown from the infrastructure. Typically, the check is only necessary of the code may run in both sync and ascnc contexts. An exception would typically unwind the stack and kill the thread, masking the source of exception and produce a segfault as a result of std::terminate().  A nice dump that would point to the site of the problem is preferable.

---
In reply to: [490640111](https://github.com/microsoft/onnxruntime/pull/5169#discussion_r490640111) [](ancestors = 490640111)

Reply:
I don't get it - do u mean letting out a segfault won't kill the thread and its attached process?
Further if REALLY we want to silence exception for whatever reason listed, invoke(...) may need addition design to ensure that like do extra check and return an bool along with ResultType to indicate the call status, a bit like std::set::emplace maybe.

---
In reply to: [491117771](https://github.com/microsoft/onnxruntime/pull/5169#discussion_r491117771) [](ancestors = 491117771,490640111)

Reply:
My point is that if the exception is caught it provides no debugging value bc the stack is gone.

---
In reply to: [491146638](https://github.com/microsoft/onnxruntime/pull/5169#discussion_r491146638) [](ancestors = 491146638,491117771,490640111)